specs/DataSet.rst

```diff
@@ -0,0 +1,253 @@
+=====================
+DataSet Specification
+=====================
+
+Introduction
+============
+
+The DataSet class is used in QCoDeS to hold measurement results.
+It is the destination for measurement loops and the source for plotting and data analysis.
+As such, it is a central component of QCoDeS.
+
+The DataSet class should be usable on its own, without other QCoDeS components.
+In particular, the DataSet class should not require the use of Loop and parameters, although it should integrate with those components seamlessly.
+This will significantly improve the modularity of QCoDeS by allowing users to plug into and extend the package in many different ways.
+As long as a DataSet is used for data storage, users can freely select the QCoDeS components they want to use.
+
+Terminology
+================
+
+Parameter
+    A logically-single value input to or produced by a measurement.
+    A parameter need not be a scalar, but can be an array or a tuple or an array of tuples, etc.
+    A DataSet parameter corresponds conceptually to a QCoDeS parameter, but does not have to be defined by or associated with a QCoDeS Parameter . 
+    Roughly, a parameter represents a column in a table of experimental data.
+    
+Result
+    A result is the collection of parameter values associated to a single measurement in an experiment.
+    Roughly, a result corresponds to a row in a table of experimental data.
+    
+Role
+    Parameters may play different roles in a measurement.
+    Specifically, they may be input to the measurement (set-points) or outputs of the measurement (measured or computed values).
+    This distinction is important for plotting and for replicating an experiment.
+    
+DataSet
+    A DataSet is a QCoDeS object that stores the results of an experiment.
+    Roughly, a DataSet corresponds to a table of experimental data, along with metadata that describes the data    .
+    Depending on the state of the experiment, a DataSet may be "in progress" or "completed".
+
+ExperimentContainer
+    An ExperimentContainer is a QCoDeS object that stores all information about an experiment.
```

Comment:
@majacassidy  thanks for the feedback. 

-  I would say yes,  although it's more of an implementation than a concept issue (i.e.  how to make so that it's easy to do this).
-  Datasets live on their onw, so yes. The container just acts like a file system on steroids (allowing to perform searches and so on. I guess we would also want to add this feature: select x from container z, and y from container w and compare them. 

- Yes. Although one is always limited by hardware that can't operate simultaneously. 
- Yes. 
- The first implementation will store a pointer to the dataset. But it may in the long run be way more convenient to store the actual data (but this opens up a lot of accessibility problems).
    The container would store:
        - references to all the datasets generated with an unique hash and a timestamp 
        - metadata of all the datasets  linked to the unique hash of the dataset they describe
        - metadata of the experiment (  something that a user would want to stick inside, literally anything) 
        - GitHub hash of the qcodes version one is using ( easy tro trace bugs, make refeneces), and a diff of what is changed locally (custom drivers and so on)
          
        -  scripts (but to do that we'd have to agree on a standard way to include them

re: parameters those are saved in the metadata of every dataset.